include/BSMPT/utility/utility.h

```diff
@@ -67,6 +108,138 @@ std::ostream &operator<<(std::ostream &os, const std::vector<T> &vec)
   return os;
 }
 
+/**
+ * @brief vector addition
+ */
+template <typename T>
+std::vector<T> operator+(const std::vector<T> &a, const std::vector<T> &b)
+{
+  assert(a.size() == b.size());
+  std::vector<T> result;
+  result.reserve(a.size());
+
+  std::transform(a.begin(),
+                 a.end(),
+                 b.begin(),
+                 std::back_inserter(result),
+                 std::plus<T>());
+  return result;
+}
+
+/**
+ * @brief vector subtraction
+ */
+template <typename T>
+std::vector<T> operator-(const std::vector<T> &a, const std::vector<T> &b)
+{
+  assert(a.size() == b.size());
+
+  std::vector<T> result;
+  result.reserve(a.size());
+
+  std::transform(a.begin(),
+                 a.end(),
+                 b.begin(),
+                 std::back_inserter(result),
+                 std::minus<T>());
+  return result;
+}
+
+/**
+ * @brief multiplication of vector with scalar
+ */
+template <typename T, typename T2>
+std::vector<T> operator*(const T2 &a, const std::vector<T> &b)
+{
+  std::vector<T> result;
+  result.reserve(b.size());
+
+  std::transform(b.begin(),
+                 b.end(),
+                 std::back_inserter(result),
+                 [&a](T i) { return a * i; });
+  return result;
+}
+
+/**
+ * @brief division of vector by scalar
+ */
+template <typename T, typename T2>
+std::vector<T> operator/(const std::vector<T> &a, const T2 &b)
+{
+  std::vector<T> result;
+  result.reserve(a.size());
+
+  std::transform(a.begin(),
+                 a.end(),
+                 std::back_inserter(result),
+                 [&b](T i) { return i / b; });
+  return result;
+}
+
+/**
+ * @brief dot product of two vectors
+ */
+template <typename T>
+T operator*(const std::vector<T> &a, const std::vector<T> &b)
+{
+  assert(a.size() == b.size());
+
+  std::vector<T> result;
+  result.reserve(a.size());
+
+  std::transform(a.begin(),
+                 a.end(),
+                 b.begin(),
+                 std::back_inserter(result),
+                 [](T i, T j) { return (i * j); });
+
+  T result1 = std::accumulate(result.begin(), result.end(), 0.0);
+
+  return result1;
+}
+
+/**
+ * @brief multiplication of matrix with vector
+ */
+template <typename T>
+std::vector<T> operator*(const std::vector<std::vector<T>> &a,
```

Comment:
Please be aware with the multiple template overloads for the `*` operator. As the matrix*vector and dot product can be catched by the multiplication with a scalar template, the order here is very important. You should add unit tests for this to ensure it does not introduce a bug during some reordering.

Reply:
Added unit tests for this. ```vector * matrix``` does not compile.